### PR TITLE
Fuzz testing

### PR DESCRIPTION
**Pull request by braydonf**
Todo:
- [x] Generate random data
- [x] Run scripts against both bitcoin-abc and bcash
- [x] Test both positive and negative test cases

Improvements:
- [ ] Speed up run time (fork or shared memory for message passing)
- [ ] Generate test cases using AFL

### CODE REVIEW

File: test-fuzz/verify.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    const inputIndex = parseInt(lines[2]);
+    const input = tx.inputs[inputIndex].script;
+    const output = bcoin.Script.fromRaw(outputScriptBuffer);
+
+    const witness = null; // segwit todo
+    const flags = parseInt(lines[3]);
+
+    bcoin.Script.verify(
+      input,
+      witness, // segwit todo
```
Comment (bucko13) [inline test-fuzz/verify.js]: won't need this for bcash implementation, just bcoin right?

Comment (braydonf) [reply test-fuzz/verify.js]: yeah, i'll need to update. i had originally run this with bcoin.

**Comment by braydonf** [comment]
Basic functionality here is complete. There are some improvements that can be made to speed it up and some further work with using AFL to generate test data.

**Comment by braydonf** [comment]
All test looking good from running against  [jonasnick's test cases](https://github.com/jonasnick/bitcoinconsensus_testcases)

There were a few things discovered in the process though. The behavior with `libbitcoinconsensus` does not represent actual script execution results as there is 'extra' validation of the `flags` for the purposes of working with the API. No issues otherwise. Here were the cases:

------------

**Test case:**
jonasnick/bitcoinconsensus_testcases/0.10-positive/03a5f65ff10ebc2ff2569d3eac29df98ae5d9b6b

**There was an execution difference at:**
https://github.com/Bitcoin-ABC/bitcoin-abc/blob/master/src/script/bitcoinconsensus.cpp#L120

**Description:**
There was a new method to the bindings added that includes a new function signature with an amount argument, that function should be used in the case that SCRIPT_ENABLE_SIGHASH_FORKID or SCRIPT_FLAGS_VERIFY_WITNESS_DEPRECATED is used. There was a problem with the test script not runnning the correct methods, and not a difference in the implementation.

-------------

**Test case:**
jonasnick/bitcoinconsensus_testcases/0.10-positive/c03e7d31e726d783043102443a16e39127ac79a2

**There was/is an execution difference at:**
https://github.com/Bitcoin-ABC/bitcoin-abc/blob/master/src/script/bitcoinconsensus.cpp#L80

**Description:**
There is "extra" validation that is not a part of the consensus code, that is validating the input to the libary functions. This validation is failing with invalid flags, however after running the test with this section disabled, the results were identical between the implementations.

-------------

File: test-fuzz/verify.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    console.error(err)
+    process.exit(2);
+  }
+
+  try {
+    const lines = data.split('\n');
+    const outputScriptBuffer = Buffer.from(lines[0], 'hex');
+    const txBuffer = Buffer.from(lines[1], 'hex');
+
+    const tx = bcash.TX.fromRaw(txBuffer);
```
Comment (bucko13) [inline test-fuzz/verify.js]: Does it make sense to pull out the modules we need from the library at the top of the file. 
I think this is what we're doing in other places:
```
const TX  = ('../lib/primitives/tx');
const Script = ('../lib/script');
```

Comment (braydonf) [reply test-fuzz/verify.js]: Updated

**Comment by braydonf** [comment]
Closing, as this can be done separately. We can bring in test cases that are found to be run with standard test suite, if necessary. Will leave the branch up for reference.

**Pull request closed without merging**